### ISSUE TITLE
ServiceFabric + ETW : behavior issue

### ISSUE THREAD

**Issue by CITnDev** [Azure/diagnostics-eventflow#96]
Hello,

Based on your documentation, the code with service fabric is :
```csharp
        using (var diagnosticsPipeline = ServiceFabricDiagnosticPipelineFactory.CreatePipeline("MyApplication-MyService-DiagnosticsPipeline"))
        {
            ServiceRuntime.RegisterServiceAsync("MyServiceType", ctx => new MyService(ctx)).Wait();

            ServiceEventSource.Current.ServiceTypeRegistered(Process.GetCurrentProcess().Id, typeof(MyService).Name);

            Thread.Sleep(Timeout.Infinite);
        }
```

I use the ETW input but each time the service starts, a new TraceEventSession is created by this code in StandardTraceEventSession.cs :
```csharp
this.inner = new TraceEventSession($"EventFlow-{nameof(EtwInput)}-{Guid.NewGuid().ToString()}", TraceEventSessionOptions.Create);
```

But when i request a stop to the service, Service Fabric engine should force the program exit : the DiagnosticPipeline Dispose seems to be never called. So the TraceEventSession stays running. After several restarts I have a system resource error because of the limit of maximum running TraceEventSession allowed by Windows.
I did a workaround with an event in my Stateless service which is triggered on OnClose and OnAbort methods to be able to call the DiagnosticPipeline Dispose method to free the running TraceEventSession.

```csharp
        using (var diagnosticsPipeline = ServiceFabricDiagnosticPipelineFactory.CreatePipeline("MyApplication-MyService-DiagnosticsPipeline"))
        {
            ServiceRuntime.RegisterServiceAsync("MyServiceType", ctx => {
                    var service = new MyService(ctx);
                    service.OnStopping += diagnosticsPipeline.Dispose();
                    return service;
            }).Wait();

            ServiceEventSource.Current.ServiceTypeRegistered(Process.GetCurrentProcess().Id, typeof(MyService).Name);

            Thread.Sleep(Timeout.Infinite);
            // Code here never called
        }
```

Is there a better way to Dispose the TraceEventSession ? Maybe it is a question for ServiceFabric team.

**Comment by karolz-ms** [comment]
@CITnDev you are absolutely right. The documentation is wrong. I am going to work with the Service Fabric team to come up with a better recommendation.

Off top of my head what you are doing makes sense (to handle pipeline disposal in OnClose and OnAbort). I just want to make sure that we cover all possible ways for the service to "end its life".

**Comment by mattrowmsft** [comment]
@karolz-ms Can we add a finalizer for the trace session manager in the ETW plugin?
@CITnDev Beware the lifecycle methods you mentioned are tied to replicas while the process would be for the deployed code package.  If you are using a partitioned stateful service, it is possible that 2 replicas for coexist in the same process.  When one replica closes it would dispose the pipeline still being used by the other replica.

**Comment by CITnDev** [comment]
Thank you Matt for your reply. 
I will test with a finalizer and if the behavior is good I ll do a pull request.

**Comment by mattrowmsft** [comment]
Karol pointed out to me that although this specific issue is related to the session, there is also larger issue of pipeline not being disposed properly. I imagine we might want both dispose and finalizer though.

**Comment by karolz-ms** [comment]
Sebastien, I am not sure if the finalizer solution is a good one. Finalizers are really designed for freeing native resources held by managed classes. As far as I know there are no guarantees CLR makes in terms of order of finalization, so if your (managed) class holds a reference to another managed class, that other class might already be finalized before your class'es finalizer is called.

I have an alternative idea involving using Console.CancelKeyPress event to ensure that the pipeline gets disposed; let me do some experiments and report the result here.

**Comment by karolz-ms** [comment]
It may be as simple as adding one more line to the Main() method implementation (right after the statement that creates the diagnostic pipeline):

```csharp
public static void Main(string[] args)
{
    try
    {
        using (var diagnosticsPipeline = ServiceFabricDiagnosticPipelineFactory.CreatePipeline("MyApplication-MyService-DiagnosticsPipeline"))
        {
            Console.CancelKeyPress += (sender, eventArgs) => diagnosticsPipeline.Dispose();

            ServiceRuntime.RegisterServiceAsync("MyServiceType", ctx => new MyService(ctx)).Wait();

            ServiceEventSource.Current.ServiceTypeRegistered(Process.GetCurrentProcess().Id, typeof(MyService).Name);

            Thread.Sleep(Timeout.Infinite);
        }
    }
    catch (Exception e)
    {
        ServiceEventSource.Current.ServiceHostInitializationFailed(e.ToString());
        throw;
    }
}
```

@CITnDev let me know if that solves the issue you are seeing

**Comment by karolz-ms** [comment]
This might be even more robust (thanks for the suggestion, @mattrowmsft !)

```csharp
    public class Program
    {
        public static void Main(string[] args)
        {
            try
            {
                ManualResetEvent terminationEvent = new ManualResetEvent(initialState: false);

                using (var diagnosticsPipeline = ServiceFabricDiagnosticPipelineFactory.CreatePipeline("DiagnosticsPipeline-FrontendService"))
                {
                    Console.CancelKeyPress += (sender, eventArgs) => Shutdown(diagnosticsPipeline, terminationEvent);

                    AppDomain.CurrentDomain.UnhandledException += (sender, unhandledExceptionArgs) =>
                    {
                        ServiceEventSource.Current.UnhandledException(unhandledExceptionArgs.ExceptionObject?.ToString() ?? "(no exception information)");
                        Shutdown(diagnosticsPipeline, terminationEvent);
                    };

                    // This is the name of the ServiceType that is registered with FabricRuntime. 
                    // This name must match the name defined in the ServiceManifest. If you change
                    // this name, please change the name of the ServiceType in the ServiceManifest.
                    ServiceRuntime.RegisterServiceAsync("AirTrafficControlWebType", ctx => new AirTrafficControlWeb(ctx)).Wait();

                    ServiceEventSource.Current.ServiceTypeRegistered(Process.GetCurrentProcess().Id, typeof(AirTrafficControlWeb).Name);

                    terminationEvent.WaitOne();
                    terminationEvent.Dispose();
                }
            }
            catch (Exception e)
            {
                ServiceEventSource.Current.ServiceHostInitializationFailed(e.ToString());
                throw;
            }
        }

        private static void Shutdown(IDisposable disposable, ManualResetEvent terminationEvent)
        {
            try
            {
                disposable.Dispose();
            }
            finally
            {
                terminationEvent.Set();
            }
        }
    }
```

The UnhandledException event is a very simple addition to the standard ServiceEventSource:

```csharp
    [Event(UnhandledExceptionEventId, Level = EventLevel.Error, Message = "An unhandled exception has occurred")]
    public void UnhandledException(string exception)
    {
       WriteEvent(UnhandledExceptionEventId, exception);
    }
```

**Comment by karolz-ms** [comment]
I have just merged a doc update with the above code snipped. Please let me know if this does not work for you.

**Comment by TheAzureGuy** [comment]
Hey guys, has anybody tested the above sample code? I doubt this is the case.

Here is some bad news...

I can confirm that as soon as you subscribe to CancelKeyPress, the SF service won't be able to exit gracefully within the allotted amount of time (59 seconds if I recall from the spec). This results in a longer than expected restart cycle and noisy warnings in the SF Explorer, similar to the following:

`There was an error during CodePackage activation.The process/container terminated with exit code:7148` 

This behavior persists even after setting .Cancel = true on the event args as suggested here:

https://dotnet.github.io/orleans-docs/Documentation/Deployment-and-Operations/SiloShutdown.html

In summary, the above approach does NOT work as advertised. I'm asking the PG folks to investigate and suggest a solution. Thank you!

**Comment by karolz-ms** [comment]
@TheAzureGuy thanks for letting us know.  Could you please open a new issue about the problem you have encountered?

Please let us know your versions of Service Fabric, .NET/.NET Core, and the OS. An example (e.g. GitHub gist) that illustrates the problem would be very helpful.

Thank again!

**Comment by TheAzureGuy** [comment]
@karolz-ms Here you go: https://github.com/microsoft/service-fabric/issues/936
